File: typescript/problem_11.ts
// TrieNode holds children and a flag marking end-of-word.
class TrieNode {
    children: Map<string, TrieNode> = new Map();
    isEnd = false;
}

// Trie supports O(|prefix|) insert and O(|prefix| + k) autocomplete.
export class Trie {
    private root = new TrieNode();

    insert(word: string): void {
        let node = this.root;
        for (const ch of word) {
            if (!node.children.has(ch)) node.children.set(ch, new TrieNode());
            node = node.children.get(ch)!;
        }
        node.isEnd = true;
    }

    private collect(node: TrieNode, prefix: string, results: string[]): void {
        if (node.isEnd) results.push(prefix);
        for (const [ch, child] of [...node.children.entries()].sort())
            this.collect(child, prefix + ch, results);
    }

    // autocomplete returns all words sharing the given prefix, sorted.
    autocomplete(prefix: string): string[] {
        let node = this.root;
        for (const ch of prefix) {
            if (!node.children.has(ch)) return [];
            node = node.children.get(ch)!;
        }
        const results: string[] = [];
        this.collect(node, prefix, results);
        return results;
    }
}

export function problem11(): void {
    console.log("Problem 11:");
    const t = new Trie();
    for (const w of ["dog", "deer", "deal"]) t.insert(w);
    const got = t.autocomplete("de");
    if (got.length !== 2 || got[0] !== "deal" || got[1] !== "deer") throw new Error("assertion failed");
    if (t.autocomplete("xyz").length !== 0) throw new Error("assertion failed");
    console.log("All Tests Passed");
}

File: rust/problem_11.rs
use std::collections::HashMap;

// TrieNode holds children and a flag marking end-of-word.
#[derive(Default)]
pub struct TrieNode {
    children: HashMap<char, TrieNode>,
    is_end: bool,
}

// Trie supports O(|prefix|) insert and O(|prefix| + k) autocomplete.
#[derive(Default)]
pub struct Trie {
    root: TrieNode,
}

impl Trie {
    pub fn new() -> Self { Self::default() }

    pub fn insert(&mut self, word: &str) {
        let mut node = &mut self.root;
        for ch in word.chars() {
            node = node.children.entry(ch).or_default();
        }
        node.is_end = true;
    }

    fn collect(node: &TrieNode, prefix: String, results: &mut Vec<String>) {
        if node.is_end { results.push(prefix.clone()); }
        let mut keys: Vec<char> = node.children.keys().cloned().collect();
        keys.sort();
        for ch in keys {
            Self::collect(&node.children[&ch], format!("{}{}", prefix, ch), results);
        }
    }

    // autocomplete returns all words sharing the given prefix, sorted.
    pub fn autocomplete(&self, prefix: &str) -> Vec<String> {
        let mut node = &self.root;
        for ch in prefix.chars() {
            match node.children.get(&ch) {
                Some(n) => node = n,
                None    => return vec![],
            }
        }
        let mut results = Vec::new();
        Self::collect(node, prefix.to_string(), &mut results);
        results
    }
}

pub fn problem11() {
    println!("Problem 11:");
    let mut t = Trie::new();
    for w in ["dog", "deer", "deal"] { t.insert(w); }
    let got = t.autocomplete("de");
    assert_eq!(got, vec!["deal", "deer"], "assertion failed");
    assert!(t.autocomplete("xyz").is_empty(), "assertion failed");
    println!("All Tests Passed");
}
